### PR TITLE
Faster determinant and Eigen Value Calculation

### PR DESCRIPTION
**Pull request by sudoWin**
Eigen Values and Determinant computation can be sped-up in the case of Lower/Upper Triangular Matrices

Currently what the algorithm does is, 
It calculates the 
1. The determinant of A-(lambda)I
2. Solves the polynomial to obtain eigen values,

But in a triangular matrix, the eigenvalues are diagonal entries so this PR avoids useless calculation and solving,

In determinant of a triangular matrix, the answer is the product of all diagonal elements, so this PR also adds that little speedup.

This issue has been taken from Sympy:
https://github.com/sympy/sympy/issues/18607

### CODE REVIEW

**Comment by sudoWin** [comment]
Please review @isuruf

**Comment by sudoWin** [comment]
Please review @isuruf Sir

File: symengine/dense_matrix.cpp [hunk 1]
```diff
@@ -1613,6 +1631,25 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
 
 RCP<const Set> eigen_values(const DenseMatrix &A)
 {
+    unsigned n = A.nrows();
+    bool is_upper = true, is_lower = true;
+    for (unsigned i = 0; i < n; ++i) {
+        for (unsigned j = 0; j < n; ++j) {
+            if (i > j and not eq(*(A.get(i, j)), *zero))
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Can you use braces?
```suggestion
            if (i > j and not eq(*(A.get(i, j)), *zero)) {
```

File: symengine/dense_matrix.cpp [hunk 2]
```diff
@@ -1613,6 +1631,25 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
...
+                is_lower = false;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Can you check that `is_upper` is also false and if so, stop the two loops completely?

File: symengine/dense_matrix.cpp (hunk 1 above)
Comment (sudoWin) [reply symengine/dense_matrix.cpp]: Okay I am sending a PR right now. Thanks for reviewing :)

File: symengine/dense_matrix.cpp [hunk 3]
```diff
@@ -1613,6 +1631,25 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
...
+            else if (i < j and not eq(*(A.get(i, j)), *zero))
+                is_upper = false;
+        }
+    }
+
+    if (is_lower or is_upper) {
+        RCP<const Set> eigenvals = emptyset();
+        for (unsigned i = 0; i < n; ++i) {
+            eigenvals = set_union({eigenvals, finiteset({A.get(i, i)})});
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: You can create one `finiteset` using all of the eigenvalues instead of creating a `finiteset` incrementally.

Comment (sudoWin) [reply symengine/dense_matrix.cpp]: How can that be done? I tried working with finite set and this seemed to be the only way I could make it work

Comment (isuruf) [reply symengine/dense_matrix.cpp]: `finiteset` takes in a C++ set `set_basic`. You can create a `set_basic` first and then pass that to `finiteset`.

Comment (sudoWin) [reply symengine/dense_matrix.cpp]: Thanks, I will push the changes in a few minutes

**Comment by sudoWin** [comment]
Although the Travis Test has passed, it is still showing yellow. Please review @isuruf

File: symengine/dense_matrix.cpp [hunk 4]
```diff
@@ -1613,6 +1632,30 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
...
+    unsigned n = A.nrows();
+    bool is_upper = true, is_lower = true;
+    for (unsigned i = 0; i < n; ++i) {
+        for (unsigned j = 0; j < n; ++j) {
+            if (i > j and not eq(*(A.get(i, j)), *zero)) {
+                is_lower = false;
+                break;
+            } else if (i < j and not eq(*(A.get(i, j)), *zero)) {
+                is_upper = false;
+                break;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Breaking here would interfere with checking `is_lower`.

Comment (isuruf) [reply symengine/dense_matrix.cpp]: I suggest making two separate checks. One for `is_lower` and one for `is_upper`

Comment (sudoWin) [reply symengine/dense_matrix.cpp]: I got it, Thanks for your guidance :)

File: symengine/dense_matrix.cpp [hunk 5]
```diff
@@ -1613,6 +1632,37 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
 
 RCP<const Set> eigen_values(const DenseMatrix &A)
 {
+    unsigned n = A.nrows();
+    bool is_upper = true, is_lower = true;
+    for (unsigned i = 0; i < n; ++i) {
+        for (unsigned j = 0; j < n; ++j) {
+            if (i > j and not eq(*(A.get(i, j)), *zero)) {
+                is_lower = false;
+                break;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: You need to break both the loops. It's easier if this was a separate helper function.

File: symengine/dense_matrix.cpp [hunk 6]
```diff
@@ -1613,6 +1632,37 @@ void cross(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
 
 RCP<const Set> eigen_values(const DenseMatrix &A)
 {
+    unsigned n = A.nrows();
+    bool is_upper = true, is_lower = true;
+    for (unsigned i = 0; i < n; ++i) {
+        for (unsigned j = 0; j < n; ++j) {
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]:
```suggestion
        for (unsigned j = 0; j < i; ++j) {
```

File: symengine/dense_matrix.cpp (hunk 5 above)
Comment (isuruf) [reply symengine/dense_matrix.cpp]: nvm

Comment (sudoWin) [reply symengine/dense_matrix.cpp]: So should we add a property is_upper and is_lower to Class DenseMatrix??

**Comment by sudoWin** [comment]
please review @isuruf

**Comment by isuruf** [comment]
@sudoWin, I'm sorry this PR fell off the radar. This looks good to me. There are two things I'd like to see.
1. Move the `is_lower` and `is_upper` to 2 helper functions.
2. Add a test

**Comment by certik** [comment]
Indeed, sorry about that @sudoWin. Please let us know if you have time to finish it.

**Comment by sudoWin** [comment]
It's ok. I would definitely like to start contributing after my exams end.

File: symengine/dense_matrix.cpp [hunk 7]
```diff
@@ -63,6 +63,44 @@ unsigned DenseMatrix::rank() const
...
+bool DenseMatrix::is_upper() const
+{
+    bool is_upper = true;
+    auto A = *this;
+    unsigned n = A.nrows();
+    for (unsigned i = 0; i < n - 1; ++i) {
+        for (unsigned j = i + 1; j < n; ++j) {
+            if (neq(*(A.get(i, j)), *zero)) {
+                is_upper = false;
+                break;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp:94]:
```suggestion
                return False
```

File: symengine/dense_matrix.cpp [hunk 8]
```diff
@@ -63,6 +63,44 @@ unsigned DenseMatrix::rank() const
...
+            }
+        }
+        if (not is_upper) {
+            break;
+        }
```
Comment (isuruf) [inline symengine/dense_matrix.cpp:99]:
```suggestion
```

**Comment by sudoWin** [comment]
Please review @isuruf @certik sir.

File: symengine/dense_matrix.cpp [hunk 9]
```diff
@@ -63,6 +63,34 @@ unsigned DenseMatrix::rank() const
     throw NotImplementedError("Not Implemented");
 }
 
+bool DenseMatrix::is_lower() const
+{
+    auto A = *this;
+    unsigned n = A.nrows();
+    for (unsigned i = 1; i < n; ++i) {
+        for (unsigned j = 0; j < i; ++j) {
+            if (neq(*(A.get(i, j)), *zero)) {
```
Comment (isuruf) [inline symengine/dense_matrix.cpp:72]:
```suggestion
            if (not is_zero(*A.get(i, j))) {
```

Comment (sudoWin) [reply symengine/dense_matrix.cpp:72]: @isuruf , there's no function is_zero, which accepts Symengine::Basic

Comment (isuruf) [reply symengine/dense_matrix.cpp:72]: Have you checked latest master? It should be there in `symengine/number.h`

Comment (sudoWin) [reply symengine/dense_matrix.cpp]: Yes, I saw.
I pulled from symenging/symengine, updated the code, and pushed back.

**Comment by sudoWin** [comment]
Please review @isuruf

**Comment by sudoWin** [comment]
Are there any other changes that need to be done? Unfortunately I don't know, why the `codecov/project` is failing, so could you please tell why it is failing, or is it alright? @isuruf

**Comment by isuruf** [comment]
Thanks for the PR @sudoWin

**Pull request merged**